hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/operation/HederaSelfDestructOperationV046.java

```diff
@@ -99,14 +99,16 @@ private ExceptionalHaltReason reasonToHalt(
         if (toBeDeleted.equals(beneficiaryAddress)) {
             return HederaExceptionalHaltReason.SELF_DESTRUCT_TO_SELF;
         }
-        if (updater.contractIsTokenTreasury(toBeDeleted)) {
-            return HederaExceptionalHaltReason.CONTRACT_IS_TREASURY;
-        }
-        if (updater.contractHasAnyBalance(toBeDeleted)) {
-            return HederaExceptionalHaltReason.TRANSACTION_REQUIRES_ZERO_TOKEN_BALANCES;
-        }
-        if (updater.contractOwnsNfts(toBeDeleted)) {
-            return HederaExceptionalHaltReason.CONTRACT_STILL_OWNS_NFTS;
+        if (updater.trackingAccounts() != null) {
+            if (updater.contractIsTokenTreasury(toBeDeleted)) {
+                return HederaExceptionalHaltReason.CONTRACT_IS_TREASURY;
+            }
+            if (updater.contractHasAnyBalance(toBeDeleted)) {
+                return HederaExceptionalHaltReason.TRANSACTION_REQUIRES_ZERO_TOKEN_BALANCES;
+            }
+            if (updater.contractOwnsNfts(toBeDeleted)) {
+                return HederaExceptionalHaltReason.CONTRACT_STILL_OWNS_NFTS;
+            }
```

Comment:
AFAICT we're only using the updaters here for read-only operations. 
In theory, we should be able to detect a static message frame by calling `frame.isStatic()` and allow the frame to read from the state without being able to write. Something similar is implemented in `HTSPrecompiledContract` in order to allow query operations to execute successfully [here](https://github.com/hashgraph/hedera-services/blob/83333db0e2a23a33ad999e7b85f7ed8d7e58bfba/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/precompile/HTSPrecompiledContract.java#L201). I think we can use the tranckingLedgers from the updater which _should_ be instantiated at this point.
This way, queries and simulations would return a more correct result.

Reply:
After discussing this with Nikolay, we realized that there's a cleaner solution --- we should add a check that the original execute() method [performs](https://github.com/hyperledger/besu/blob/750580dcca349d22d024cc14a8171b2fa74b505a/evm/src/main/java/org/hyperledger/besu/evm/operation/SelfDestructOperation.java#L51) but is missing from our implementation

Reply:
:rocket: